src/mame/tandberg/tdv2100_disp_logic.cpp

```diff
@@ -0,0 +1,1063 @@
+
+	for(int char_row_nr = char_min_row; char_row_nr <= char_max_row; char_row_nr++)
+	{
+		// This is at the start of a char row
+		int const chr_y_pos = char_row_nr*14;
+		if(char_row_nr != char_min_row)
+		{
+			attribute = get_attribute(char_row_nr, char_min_col, attribute, char_row_nr-1, char_max_col+1);
+		}
+
+		for(int char_col_nr = char_min_col; char_col_nr <= char_max_col; char_col_nr++)
+		{
+			// This is at the start of a single character
+			int const chr_x_pos = char_col_nr*9;
+			int const in_line_char_nr = char_col_nr - 4;
+			int const vram_address = get_ram_addr(char_row_nr, in_line_char_nr);
+			bool const is_cursor = (m_cursor_col == in_line_char_nr && m_cursor_row == char_row_nr);
+
+			for(int line = 0; line < 14; line++)
+			{
+				// This is at the start of a 9-dot raster-line segment in a single character
+				uint8_t data = 0x00;
+				int dot_nr8 = 0;
+				int extra_intensity = 1;
+				if(in_line_char_nr >= 0 && in_line_char_nr < 80)
+				{
+					bool const draw_cursor = (is_cursor && (line == 12 || block_cursor) && (!blinking_cursor || cursor_blink_strobe));
+					int const vram_data = m_vram[vram_address];
+
+					// Pattern generator & character blanking
+					if(m_video_enable && (vram_data&0x60 || !blank_ctrl_chars) && !(vram_data&0x80 && blank_attr_chars))
+					{
+						data = m_font[((vram_data&0x7f)<<4) + line];
+						if(extend_dot7)
+						{
+							dot_nr8 = data&0x01;
+						}
+					}
+
+					// Pattern modifiers
+					if(underline_mode)
+					{
+						// Note: The underline in underline-mode is not affected by blanking!
+						if((vram_data&0x80) && line == 13)
+						{
+							data = 0xff;
+							dot_nr8 = 0x01;
+						}
+					}
+					else if(m_video_enable) // attribute mode
+					{
+						if(vram_data&0x80)
+						{
+							attribute = (vram_data&0x70)>>4;
+						}
+						else
+						{
+							if(attribute == 2)
+							{
+								// Low intensity
+								extra_intensity = 0;
+							}
+							else if(attribute == 3 && blink_strobe && !draw_cursor)
+							{
+								// Blink
+								data = 0x00;
+								dot_nr8 = 0x00;
+							}
+							else if(attribute == 4)
+							{
+								// Inverse
+								data = ~data;
+								dot_nr8 = ~dot_nr8;
+							}
+							else if(attribute == 5 && line == 13)
+							{
+								// Underline
+								data = 0xff;
+								dot_nr8 = 0x01;
+							}
+							else if(attribute == 6)
+							{
+								// Invissible
+								data = 0x00;
+								dot_nr8 = 0x00;
+							}
+						}
+					}
+
+					if(draw_cursor)
+					{
+						data = ~data;
+						dot_nr8 = ~dot_nr8;
+					}
+				}
+				if(full_inverse_video)
+				{
+					data = ~data;
+					dot_nr8 = ~dot_nr8;
+				}
+				for(int dot = 0; dot < 8; dot++)
+				{
+					bitmap.pix(chr_y_pos + line, chr_x_pos + dot) = palette[BIT(data, 7 - dot)<<extra_intensity];
+				}
+				bitmap.pix(chr_y_pos + line, chr_x_pos + 8) = palette[BIT(dot_nr8, 0)<<extra_intensity];
```

Comment:
You’re better off avoiding the repeated calls to `bitmap.pix` here – the compiler may or may not optimise away the repeated line address calculation.  You can lift the invariant with something like:
```c++
				auto *pix = &bitmap.pix(chr_y_pos + line, chr_x_pos);
				for(int dot = 0; dot < 8; dot++)
				{
					*pix++ = palette[BIT(data, 7 - dot)<<extra_intensity];
				}
				*pix = palette[BIT(dot_nr8, 0)<<extra_intensity];
```

Alternatively, you could index the pixel pointer rather than incrementing it.

Reply:
Changed, although that sort of optimization does sound a little extreme to me. Then again, I am no expert about the inner workings of C++ compilers.